scipy/cluster/hierarchy.py

```diff
@@ -703,12 +714,23 @@ def linkage(y, method='single', metric='euclidean'):
 
     n = int(distance.num_obs_y(y))
     method_code = _LINKAGE_METHODS[method]
+
     if method == 'single':
-        return _hierarchy.mst_single_linkage(y, n)
+        if optimal_ordering:
+            return optimal_leaf_ordering(_hierarchy.mst_single_linkage(y, n), y)
+        else:
+            return _hierarchy.mst_single_linkage(y, n)
     elif method in ['complete', 'average', 'weighted', 'ward']:
-        return _hierarchy.nn_chain(y, n, method_code)
+        if optimal_ordering:
+            return optimal_leaf_ordering(_hierarchy.nn_chain(y, n, method_code), y)
+        else:
+            return _hierarchy.nn_chain(y, n, method_code)
     else:
-        return _hierarchy.fast_linkage(y, n, method_code)
+        if optimal_ordering:
+            return optimal_leaf_ordering(_hierarchy.fast_linkage(y, n, method_code), y)
+        else:
+            return _hierarchy.fast_linkage(y, n, method_code)
```

Comment:
This would look better if the ``if method`` and ``if optimal_ordering`` statements were not interleaved. So you could change ``return _hierarchy.fast_linkage(y, n, method_code)`` to ``result = _hierarchy.fast_linkage(y, n, method_code)`` and then below the ``else`` from ``if method`` the ``if optimal_ordering:`` statement.